File: ios-native/Sources/CoreNetwork/AuthAPI.swift
import Foundation

/// Endpoints that do not need — and must not wait on — an access token.
///
/// Served by the bare ``HTTPClient``, so a failing refresh cannot recurse into
/// itself through ``TokenRefresher``.
///
/// Every path is **relative**: a leading slash would read as an absolute path
/// and would erase the `/app` prefix that production and qualif are served
/// under. `APIURLTests` pins this.
public struct AuthAPI: Sendable {

    private let client: HTTPClient

    public init(client: HTTPClient) {
        self.client = client
    }

    public func login(_ body: TodoLoginRequest) async throws(APIError) -> TodoAuthResult {
        try await client.send(
            .init(.post, "api/v1/auth/login", body: JSONBody.encode(body)),
            as: TodoAuthResult.self
        )
    }

    public func register(_ body: TodoRegisterRequest) async throws(APIError) -> TodoAuthResult {
        try await client.send(
            .init(.post, "api/v1/auth/register", body: JSONBody.encode(body)),
            as: TodoAuthResult.self
        )
    }

    public func refresh(_ body: TodoRefreshRequest) async throws(APIError) -> TodoAuthResult {
        try await client.send(
            .init(.post, "api/v1/auth/refresh", body: JSONBody.encode(body)),
            as: TodoAuthResult.self
        )
    }

    public func logout(_ body: TodoRefreshRequest) async throws(APIError) {
        try await client.send(.init(.post, "api/v1/auth/logout", body: JSONBody.encode(body)))
    }

    /// Finishes a mindlog id sign-in whose account carried no email address.
    public func completeMindlogId(
        _ body: TodoCompleteMindlogIdRequest
    ) async throws(APIError) -> TodoAuthResult {
        try await client.send(
            .init(.post, "api/v1/auth/mindlog-id/complete", body: JSONBody.encode(body)),
            as: TodoAuthResult.self
        )
    }

    /// Public. `authProviders` is what tells the sign-in screen whether to offer
    /// the mindlog id button at all — offering one this deployment has not
    /// configured just produces a 503 under the user's finger.
    public func version() async throws(APIError) -> TodoVersionInfo {
        try await client.send(.init(.get, "api/v1/version"), as: TodoVersionInfo.self)
    }
}

File: android-native/core/network/src/main/kotlin/today/mindlog/todo/core/network/api/AuthApi.kt
package today.mindlog.todo.core.network.api

import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import today.mindlog.todo.core.network.model.AuthResult
import today.mindlog.todo.core.network.model.CompleteMindlogIdRequest
import today.mindlog.todo.core.network.model.LoginRequest
import today.mindlog.todo.core.network.model.RefreshRequest
import today.mindlog.todo.core.network.model.RegisterRequest
import today.mindlog.todo.core.network.model.VersionInfo

/**
 * Endpoints that do not need — and must not wait on — an access token.
 *
 * Served by the bare OkHttp client, so a failing refresh cannot recurse into
 * itself through [today.mindlog.todo.core.network.TokenAuthenticator].
 *
 * Note every path is **relative**: a leading slash would be an absolute path
 * and would erase the `/app` prefix that production and qualif are served
 * under. `ApiUrlTest` pins this.
 */
interface AuthApi {

    @POST("api/v1/auth/login")
    suspend fun login(@Body body: LoginRequest): AuthResult

    @POST("api/v1/auth/register")
    suspend fun register(@Body body: RegisterRequest): AuthResult

    @POST("api/v1/auth/refresh")
    suspend fun refresh(@Body body: RefreshRequest): AuthResult

    @POST("api/v1/auth/logout")
    suspend fun logout(@Body body: RefreshRequest)

    /** Finishes a mindlog id sign-in whose account carried no email address. */
    @POST("api/v1/auth/mindlog-id/complete")
    suspend fun completeMindlogId(@Body body: CompleteMindlogIdRequest): AuthResult

    /**
     * Public. `authProviders` is what tells the sign-in screen whether to offer
     * the mindlog id button at all — offering one this deployment has not
     * configured just produces a 503 under the user's finger.
     */
    @GET("api/v1/version")
    suspend fun version(): VersionInfo
}
